crates/oxc_sourcemap/src/decode.rs

```diff
@@ -0,0 +1,397 @@
+use crate::error::{Error, Result};
+use crate::{SourceMap, Token};
+
+#[derive(serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct JSONSourceMap {
+    file: Option<String>,
+    mappings: String,
+    sources: Vec<String>,
+    sources_content: Option<Vec<String>>,
+    names: Vec<String>,
+}
+
+pub fn decode(value: &str) -> Result<SourceMap> {
+    let json: JSONSourceMap = serde_json::from_str(value)?;
+    let file = json.file.map(Into::into);
+    let names = json.names.into_iter().map(Into::into).collect::<Vec<_>>();
+    let sources = json.sources.into_iter().map(Into::into).collect::<Vec<_>>();
+    let source_contents =
+        json.sources_content.map(|v| v.into_iter().map(Into::into).collect::<Vec<_>>());
+    let tokens = decode_mapping(&json.mappings, names.len(), sources.len())?;
+    Ok(SourceMap::new(file, names, sources, source_contents, tokens, None))
+}
+
+#[allow(clippy::cast_possible_truncation)]
+fn decode_mapping(mapping: &str, names_len: usize, sources_len: usize) -> Result<Vec<Token>> {
+    let mut tokens = vec![];
+
+    let mut dst_col;
+    let mut src_id = 0;
+    let mut src_line = 0;
+    let mut src_col = 0;
+    let mut name_id = 0;
+    let mut nums = Vec::with_capacity(6);
+
+    for (dst_line, line) in mapping.split(';').enumerate() {
+        if line.is_empty() {
+            continue;
+        }
+
+        dst_col = 0;
+
+        for segment in line.split(',') {
+            if segment.is_empty() {
+                continue;
+            }
+
+            nums.clear();
+            parse_vlq_segment_into(segment, &mut nums)?;
+            dst_col = (i64::from(dst_col) + nums[0]) as u32;
+
+            let mut src = !0;
+            let mut name = !0;
+
+            if nums.len() > 1 {
+                if nums.len() != 4 && nums.len() != 5 {
+                    return Err(Error::BadSegmentSize(nums.len() as u32));
+                }
+                src_id = (i64::from(src_id) + nums[1]) as u32;
+                if src_id >= sources_len as u32 {
+                    return Err(Error::BadSourceReference(src_id));
+                }
+
+                src = src_id;
+                src_line = (i64::from(src_line) + nums[2]) as u32;
+                src_col = (i64::from(src_col) + nums[3]) as u32;
+
+                if nums.len() > 4 {
+                    name_id = (i64::from(name_id) + nums[4]) as u32;
+                    if name_id >= names_len as u32 {
+                        return Err(Error::BadNameReference(name_id));
+                    }
+                    name = name_id;
+                }
+            }
+
+            tokens.push(Token::new(
+                dst_line as u32,
+                dst_col,
+                src_line,
+                src_col,
+                if src == !0 { None } else { Some(src) },
+                if name == !0 { None } else { Some(name) },
+            ));
+        }
+    }
+
+    Ok(tokens)
+}
+
+const B64: [i8; 256] = [
```

Comment:
Can you format this by hand and add a rustfmt skip?

Reply:
Done